integrations/langchain-js/src/BraintrustCallbackHandler.test.ts

```diff
@@ -0,0 +1,813 @@
+        logs.push((await request.json()) as LogsRequest);
+        return HttpResponse.json(["tool-span-id"]);
+      }),
+    );
+
+    const llm = new ChatOpenAI({
+      model: "gpt-4o-mini",
+    });
+
+    const calculatorSchema = z.object({
+      operation: z
+        .enum(["add", "subtract", "multiply", "divide"])
+        .describe("The type of operation to execute."),
+      number1: z.number().describe("The first number to operate on."),
+      number2: z.number().describe("The second number to operate on."),
+    });
+
+    const calculatorTool = tool(
+      // @ts-expect-error
+      ({ operation, number1, number2 }) => {
+        // Functions must return strings
+        if (operation === "add") {
+          return `${number1 + number2}`;
+        } else if (operation === "subtract") {
+          return `${number1 - number2}`;
+        } else if (operation === "multiply") {
+          return `${number1 * number2}`;
+        } else if (operation === "divide") {
+          return `${number1 / number2}`;
+        } else {
+          throw new Error("Invalid operation.");
+        }
+      },
+      {
+        name: "calculator",
+        description: "Can perform mathematical operations.",
+        schema: calculatorSchema,
+      },
+    );
+
+    const llmWithTools = llm.bindTools([calculatorTool]);
+
+    await llmWithTools.invoke("What is 3 * 12", {
+      callbacks: [handler],
+    });
+
+    await flush();
+
+    const { spans, root_span_id } = logsToSpans(logs);
+
+    expect(spans).toMatchObject([
+      {
+        span_id: root_span_id,
+        root_span_id,
+        span_attributes: {
+          name: "ChatOpenAI",
+          type: "llm",
+        },
+        input: [
+          {
+            content: "What is 3 * 12",
+            role: "user",
+          },
+        ],
+        metadata: {
+          tags: [],
+          model: "gpt-4o-mini",
+          temperature: 1,
+          top_p: 1,
+          frequency_penalty: 0,
+          presence_penalty: 0,
+          n: 1,
+          tools: [
+            {
+              type: "function",
+              function: {
+                description: "Can perform mathematical operations.",
+                name: "calculator",
+                parameters: zodToJsonSchema(calculatorSchema),
+              },
+            },
+          ],
+        },
+        output: [
+          {
+            content: "",
+            role: "assistant",
+            tool_calls: [
+              {
+                name: "calculator",
+                args: {
+                  operation: "multiply",
+                  number1: 3,
+                  number2: 12,
+                },
+                type: "tool_call",
+                id: "call_G2Qd8HzTMyFUiMafz5H4fBIi",
+              },
+            ],
+          },
+        ],
+      },
+    ]);
+  });
+
+  it("should handle parallel runnable execution", async () => {
```

Comment:
This is a good start, but I am thinking I'm not stress testing this enough. I saw an early look of a "gotcha" with the graph case where a direct LLM invoke within a node caused a span conflict (same runId). Not yet sure if that was expected or not. Doing a live check I didn't see a problem to proceed with a warning.